applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier.c

```diff
@@ -0,0 +1,763 @@
+		[LWM2M_CARRIER_ERROR_NO_ERROR] =
+			"No error",
+		[LWM2M_CARRIER_ERROR_BOOTSTRAP] =
+			"Bootstrap error",
+		[LWM2M_CARRIER_ERROR_LTE_LINK_UP_FAIL] =
+			"Failed to connect to the LTE network",
+		[LWM2M_CARRIER_ERROR_LTE_LINK_DOWN_FAIL] =
+			"Failed to disconnect from the LTE network",
+		[LWM2M_CARRIER_ERROR_FOTA_PKG] =
+			"Package refused from modem",
+		[LWM2M_CARRIER_ERROR_FOTA_PROTO] =
+			"Protocol error",
+		[LWM2M_CARRIER_ERROR_FOTA_CONN] =
+			"Connection to remote server failed",
+		[LWM2M_CARRIER_ERROR_FOTA_CONN_LOST] =
+			"Connection to remote server lost",
+		[LWM2M_CARRIER_ERROR_FOTA_FAIL] =
+			"Modem firmware update failed",
+		[LWM2M_CARRIER_ERROR_CONFIGURATION] =
+			"Illegal object configuration detected",
+		[LWM2M_CARRIER_ERROR_INIT] =
+			"Initialization failure",
+		[LWM2M_CARRIER_ERROR_INTERNAL] =
+			"Internal failure",
+		[LWM2M_CARRIER_ERROR_RUN] =
+			"Configuration failure",
+	};
+
+	__ASSERT(PART_OF_ARRAY(strerr[err->type]),
+		 "Unhandled liblwm2m_carrier error");
+
+	LOG_ERR("LWM2M_CARRIER_EVENT_ERROR: %s, reason %d", strerr[err->type], err->value);
+}
+
+void print_deferred(const lwm2m_carrier_event_t *evt)
+{
+	const lwm2m_carrier_event_deferred_t *def = evt->data.deferred;
+
+	static const char * const strdef[] = {
+		[LWM2M_CARRIER_DEFERRED_NO_REASON] =
+			"No reason given",
+		[LWM2M_CARRIER_DEFERRED_PDN_ACTIVATE] =
+			"Failed to activate PDN",
+		[LWM2M_CARRIER_DEFERRED_BOOTSTRAP_NO_ROUTE] =
+			"No route to bootstrap server",
+		[LWM2M_CARRIER_DEFERRED_BOOTSTRAP_CONNECT] =
+			"Failed to connect to bootstrap server",
+		[LWM2M_CARRIER_DEFERRED_BOOTSTRAP_SEQUENCE] =
+			"Bootstrap sequence not completed",
+		[LWM2M_CARRIER_DEFERRED_SERVER_NO_ROUTE] =
+			"No route to server",
+		[LWM2M_CARRIER_DEFERRED_SERVER_CONNECT] =
+			"Failed to connect to server",
+		[LWM2M_CARRIER_DEFERRED_SERVER_REGISTRATION] =
+			"Server registration sequence not completed",
+		[LWM2M_CARRIER_DEFERRED_SERVICE_UNAVAILABLE] =
+			"Server in maintenance mode",
+		[LWM2M_CARRIER_DEFERRED_SIM_MSISDN] =
+			"Waiting for SIM MSISDN",
+	};
+
+	__ASSERT(PART_OF_ARRAY(strdef[def->reason]),
+		 "Unhandled deferred reason");
+
+	LOG_INF("LWM2M_CARRIER_EVENT_DEFERRED: reason %s, timeout %d seconds",
+		strdef[def->reason], def->timeout);
+}
+
+int lwm2m_carrier_event_handler(const lwm2m_carrier_event_t *event)
+{
+	int err = 0;
+
+	switch (event->type) {
+	case LWM2M_CARRIER_EVENT_INIT:
+		LOG_DBG("LWM2M_CARRIER_EVENT_INIT");
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_LINK_UP:
+		LOG_DBG("LWM2M_CARRIER_EVENT_LTE_LINK_UP");
+		/* AT+CFUN=1 to be issued. */
+		err = -1;
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_LINK_DOWN:
+		LOG_DBG("LWM2M_CARRIER_EVENT_LTE_LINK_DOWN");
+		/* AT+CFUN=4 to be issued. */
+		err = -1;
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_POWER_OFF:
+		LOG_DBG("LWM2M_CARRIER_EVENT_LTE_POWER_OFF");
+		/* TODO: defer setting the modem to minimum funtional mode. */
+		err = nrf_modem_at_printf("AT+CFUN=0");
```

Comment:
Due to the fact that LwM2M carrier library takes charge of shutting down the modem, here we cannot defer issuing AT+CFUN=0, as otherwise things might break. This will be solved in the future (please refer to NCSDK-17438).